include/experimental/__p2630_bits/submdspan_extents.hpp

```diff
@@ -0,0 +1,315 @@
+// knowledge
+template <class Arg0, class Arg1> struct StaticExtentFromStridedRange {
+  constexpr static size_t value = dynamic_extent;
+};
+
+template <class Integral0, Integral0 val0, class Integral1, Integral1 val1>
+struct StaticExtentFromStridedRange<std::integral_constant<Integral0, val0>,
+                                    std::integral_constant<Integral1, val1>> {
+  constexpr static size_t value = val0 > 0 ? 1 + (val0 - 1) / val1 : 0;
+};
+
+// creates new extents through recursive calls to next_extent member function
+// next_extent has different overloads for different types of stride specifiers
+template <size_t K, class Extents, size_t... NewExtents>
+struct extents_constructor {
+  MDSPAN_TEMPLATE_REQUIRES(
+    class Slice, class... SliceSpecifiers,
+    /* requires */(!is_convertible_v<Slice, size_t> &&
+                   !is_strided_index_range<Slice>::value)
+  )
+  MDSPAN_INLINE_FUNCTION
+  constexpr static auto next_extent(const Extents &ext, const Slice &sl,
+                                    SliceSpecifiers... slices) {
+    constexpr size_t new_static_extent = StaticExtentFromRange<
+        decltype(first_of(std::declval<Slice>())),
+        decltype(last_of(integral_constant<size_t, Extents::rank() - K>(),
+                         std::declval<Extents>(),
+                         std::declval<Slice>()))>::value;
+
+    using next_t =
+        extents_constructor<K - 1, Extents, NewExtents..., new_static_extent>;
+    using index_t = typename Extents::index_type;
+    return next_t::next_extent(
+        ext, slices...,
+        index_t(last_of(integral_constant<size_t, Extents::rank() - K>(), ext,
+                        sl)) -
+            index_t(first_of(sl)));
+  }
+
+  MDSPAN_TEMPLATE_REQUIRES(
+    class Slice, class... SliceSpecifiers,
+    /* requires */ (is_convertible_v<Slice, size_t>)
+  )
+  MDSPAN_INLINE_FUNCTION
+  constexpr static auto next_extent(const Extents &ext, const Slice &,
+                                    SliceSpecifiers... slices) {
+    using next_t = extents_constructor<K - 1, Extents, NewExtents...>;
+    return next_t::next_extent(ext, slices...);
+  }
+
+  template <class OffsetType, class ExtentType, class StrideType,
+            class... SliceSpecifiers>
+  MDSPAN_INLINE_FUNCTION
+  constexpr static auto
+  next_extent(const Extents &ext,
+              const strided_index_range<OffsetType, ExtentType, StrideType> &r,
+              SliceSpecifiers... slices) {
+    using index_t = typename Extents::index_type;
+    using new_static_extent_t =
+        StaticExtentFromStridedRange<ExtentType, StrideType>;
+    if constexpr (new_static_extent_t::value == dynamic_extent) {
+      using next_t =
+          extents_constructor<K - 1, Extents, NewExtents..., dynamic_extent>;
+      return next_t::next_extent(
+          ext, slices...,
+          r.extent > 0 ? 1 + divide<index_t>(r.extent - 1, r.stride) : 0);
```

Comment:
1. Adding 1 to `std::integral_constant<size_t, Value>` results in `size_t`, so this wouldn't preserve `integral_constant`-ness.
2. Even if it resulted in `integral_constant`, the result of the ternary operator would be `common_type_t<integral_constant<size_t, Value>, int>`, which is `int` -- again, not preserving `integral_constant`-ness.
  
https://godbolt.org/z/qWEdn5xM6

```c++
#include <utility>
#include <type_traits>

template<size_t Val1, size_t Val2>
constexpr auto divide(std::integral_constant<size_t, Val1> x,
  std::integral_constant<size_t, Val2> y) {
  return std::integral_constant<size_t, Val1 / Val2>{};
}

template<size_t Value>  
using Size = std::integral_constant<size_t, Value>;

int main() {
  constexpr auto z = divide(Size<8>{}, Size<2>{});
  static_assert(std::is_same_v<std::remove_const_t<decltype(z)>, Size<4>>);

  constexpr auto w = 1 + divide(Size<8>{}, Size<2>{});
  static_assert(not std::is_same_v<std::remove_const_t<decltype(w)>, Size<4>>);
  static_assert(std::is_same_v<std::remove_const_t<decltype(w)>, size_t>);

  static_assert(not std::is_same_v<std::common_type_t<Size<4>, int>, Size<4>>);
  static_assert(not std::is_same_v<std::common_type_t<Size<4>, int>, size_t>);
  static_assert(std::is_same_v<std::common_type_t<Size<4>, int>, int>);

  return 0;
}
```

Reply:
This doesn't matter since it doesn't determine the type. Note the values I am generating here are used as constructor arguments for an extents type which is independently determined. In case static extents are generated or preserved, these values will at best be checked against the static_extents in the extents constructor.

Reply:
`integral_constant<T, Value>` implicitly converts to `T`.  This means that `1 + integral_constant<size_t, X>{} / integral_constant<size_t, Y>{}` just returns `1 + X / Y`, so you don't even need `divide` here.

https://godbolt.org/z/x48bnaW8e